### PR TITLE
Implement generic singular value penalty

### PR DESCRIPTION
**Pull request by marcusvaltonen**
This defines a large class of regularizers, and allows for several non-convex singular value penalties, both separable and non-separable. Final PR for #52.

### CODE REVIEW

File: pyproximal/proximal/SingularValuePenalty.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        \mathcal{R}_f(\mathbf{X}) = f(\boldsymbol\lambda)
+
+    where :math:`\mathbf{X}` is a matrix of size :math:`M \times N` and
+    :math:`\boldsymbol\lambda` is the corresponding singular value vector.
+
+    Parameters
+    ----------
+    dim : :obj:`tuple`
+        Size of matrix :math:`\mathbf{X}`.
+    penalty : :class:`pylops.ProxOperator`
```
Comment (mrava87) [inline pyproximal/proximal/SingularValuePenalty.py:23]: do you mean `pyproximal.ProxOperator`?

**Review by mrava87 — COMMENTED** [review]
This looks great, I like the way you have created a single operator that can use interchangeably different operator norms on the singular values.

I have very few minor comments, then I think this is ready to go!

PS: I guess now that you have handled all the points raised in the issue, it would be great to have a tutorial on using these new set of operators and how they compare on a task of your choice :D

File: pyproximal/proximal/SingularValuePenalty.py (hunk 1 above)
Comment (marcusvaltonen) [reply pyproximal/proximal/SingularValuePenalty.py:23]: Yes, will change it

File: pyproximal/proximal/SingularValuePenalty.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        Function acting on the singular values.
+
+    Notes
+    -----
+    The pyproximal implementation allows ``penalty`` to be any function on the singular
```
[the remark this answers is not in the document]
Comment (marcusvaltonen) [reply pyproximal/proximal/SingularValuePenalty.py:28]: Changing to "acting on". Yes, of course, I mean a function in a mathematical sense, not in the Python sense.

**Review by marcusvaltonen — COMMENTED** [review]
I agree with you.. Will make updates.

**Comment by marcusvaltonen** [comment]
Regarding the tutorial: I have been thinking about it, but I will add it separately, since I want it to tackle a real-world problem where you see the benefit of non-convex penalties compared to the nuclear norm penalty. However, in doing so, I will need to define some helper methods etc, hence another PR.

**Comment by mrava87** [comment]
Perfect! And yes, I didn’t mean that you should put a tutorial here, a separate PR is best :)

**Pull request merged**